tfhe/src/shortint/server_key/mod.rs

```diff
@@ -778,24 +694,77 @@ impl ServerKey {
     where
         F: Fn(u64, u64) -> u64,
     {
-        if self
-            .is_functional_bivariate_pbs_possible(ct_left, ct_right)
-            .is_err()
-        {
-            // We don't have enough space in carries, so clear them
-            self.message_extract_assign(ct_left);
-            self.message_extract_assign(ct_right);
+        let ScalingOperation {
+            order,
+            scaled_behavior,
+            unscaled_bootstrapped,
+            scale,
+        } = self
+            .get_best_bivariate_scaling(ct_left, ct_right)
+            .expect("Current parameters can't be used to make a bivariate function evaluation");
+
+        let (ct_to_scale, unscaled_ct) = match order {
+            Order::ScaleLeft => (ct_left, ct_right),
+            Order::ScaleRight => (ct_right, ct_left),
+        };
+
+        if unscaled_bootstrapped {
+            self.message_extract_assign(unscaled_ct);
         }
-        self.is_functional_bivariate_pbs_possible(ct_left, ct_right)
-            .unwrap();
 
-        let factor = MessageModulus(ct_right.degree.get() + 1);
+        let scaled = match scaled_behavior {
+            ScaledBehavior::Scaled => self.unchecked_scalar_mul(ct_to_scale, scale),
+            ScaledBehavior::BootstrappedThenScaled => {
+                self.message_extract_assign(ct_to_scale);
+
+                self.unchecked_scalar_mul(ct_to_scale, scale)
+            }
+            ScaledBehavior::ScaledInBootstrap => {
+                let lookup_table = self
+                    .generate_lookup_table(|a| (a % self.message_modulus.0 as u64) * scale as u64);
+
+                self.apply_lookup_table(ct_to_scale, &lookup_table)
```

Comment:
Why are these 2 separated ? since they both involve a PBS, isn't it simpler to just have the later case, where the scaling happens in the bootstrap

Reply:
The first case cleans the input which can allow other operations on the same input to do less PBS

The first case will be preferred (if possible) as it has as many PBS but cleans one more input

Reply:
I don't get it 🤔 
```rust
            ScaledBehavior::BootstrappedThenScaled => {
                self.message_extract_assign(ct_to_scale);

                self.unchecked_scalar_mul(ct_to_scale, scale)
            }
```
This does a PBS, then a scalar mul so the noise of the resulting scaled ct is NOMINAL * scale
```rust
  ScaledBehavior::ScaledInBootstrap => {
                let lookup_table = self
                    .generate_lookup_table(|a| (a % self.message_modulus.0 as u64) * scale as u64);

                self.apply_lookup_table(ct_to_scale, &lookup_table)
```
This does 1 PBS, which results in a scaled ciphertext with noise set to NOMINAL

Reply:
yep, the second case did not clean the input ciphertext however as it as only the scaled variant result

Reply:
Ok I see, the input is modified